### ISSUE TITLE
[Question] How do we query to filter based on 2+ array items on the front-end?

### ISSUE THREAD

**Issue by kshitizshankar** [feathersjs-ecosystem/feathers-vuex#482]
Considering a simple use case where I have the following records of Books loaded into Vuex store of the Books Service. Each book contains a list of `tags`.

Book 1 `Coding` `Javascript` `Education`
Book 2 `Javascript` `Coding` `FeathersJS`
Book 3 `Javascript` `FeathersVuex` `FeathersJS` `Coding`

How do I query for books that have `Coding` & `FeathersJS` tags?

```
query: {
    tag: {
         $in: ['Coding', 'FeathersJS']
    }
}
```
This gives results for the books that have the `Coding` or `FeathersJS` tags as expected. (Book1, Book2, Book3)

I am not sure how to query for books that contain both the tags.

On the backend with mongoose, I can use the `$all` operator to achieve this - but how do I go about doing this on the frontend? 

Currently, my workaround was to have a custom getter on the service which executes the `$in` query and then loops through the results to find the items where both tags are present.

Is there a better way to do this? Did I miss something w.r.t querying?

**Comment by J3m5** [comment]
MongoDB allows to query a array with exact equality but it that case it respect the specific order of the items in the array.

Did you try this? 
```js
query: {
    tag: ['Coding', 'FeathersJS']
}
```
Or you can use the $all operator like that:

```js
query: {
    tag: { $all: ['Coding', 'FeathersJS'] }
}
```
but you might need to [whitelist](https://docs.feathersjs.com/api/databases/common.html#options) the `$all` operator at the level service.

**Comment by kshitizshankar** [comment]
@J3m5 Thanks a lot!! I just stumbled on it as well. Ended up digging around in the code, found out that sift is being used to filter the results and the `whitelist` option is sent as a parameter.  

`$all` seems to work as expected after whitelisting it! :)

Can I safely whitelist `$and` or other mongodb operators as well? On the [mongodb $all documentation page](https://docs.mongodb.com/manual/reference/operator/query/all/) it says that the following queries are equivalent -
`{ tags: { $all: [ "ssl" , "security" ] } }` 
`{ $and: [ { tags: "ssl" }, { tags: "security" } ] }`

However, even after whitelisting both, the results are not the same when passed to the find getter ($all works as expected but $and just returns everything without any filtering). Is that expected?

**Comment by J3m5** [comment]
> Can I safely whitelist $and or other mongodb operators as well?

It depends on what you want your users to be able to do with your data, but for $and I would say yes, but there might be a risk of data leak with the $populate operator for example.

> However, even after whitelisting both, the results are not the same when passed to the find getter ($all works as expected but $and just returns everything without any filtering). Is that expected?

You should try [this](https://github.com/crcn/sift.js/issues/161)

**Comment by kshitizshankar** [comment]
If I add it to the `additionalOperators=['$elemMatch']` in the 'service-module.getters.js' file and replace it with `const additionalOperators = ['$elemMatch','$and']` - It starts working and let's me execute the `{ $and: [ { tags: "ssl" }, { tags: "security" } ] }` query.

```
const FILTERS = ['$sort', '$limit', '$skip', '$select']
const OPERATORS = ['$in', '$nin', '$lt', '$lte', '$gt', '$gte', '$ne', '$or']
const additionalOperators = ['$elemMatch']
const defaultOps = FILTERS.concat(OPERATORS).concat(additionalOperators)

export default function makeServiceGetters() {
  return {
    list(state) {
      return state.ids.map(id => state.keyedById[id])
    },
    find: state => params => {
      params = { ...params } || {}

      // Set params.temps to true to include the tempsById records
      params.temps = params.hasOwnProperty('temps') ? params.temps : false

      const { paramsForServer, whitelist } = state
      const q = _omit(params.query || {}, paramsForServer)
      const customOperators = Object.keys(q).filter(
        k => k[0] === '$' && !defaultOps.includes(k)
      )
      const cleanQuery = _omit(q, customOperators)

      const { query, filters } = filterQuery(cleanQuery, {
        operators: additionalOperators.concat(whitelist)
      })
```
Why do the top-level `$` operators get filtered out irrespective of the whitelist? Trying to get a better understanding. Not sure if there is any documentation/posts/issues that go into this..



> It depends on what you want your users to be able to do with your data, but for $and I would say yes, but there might be a risk of data leak with the $populate operator for example.

Oh ok - Though I am not using any $populate, but it sounds like something that might cause future problems. Will try and simplify my stories for now and get back to this once I have a better understanding I guess.

**Comment by J3m5** [comment]
I think it's because of this:
```js
const customOperators = Object.keys(q).filter(
  k => k[0] === '$' && !defaultOps.includes(k)
)
const cleanQuery = _omit(q, customOperators)
```

The filtering doesn't take the whitelist into account, I've seen it before but I don't know why it's done because the `cleanQuery` function is there for that and it is recursive.

**Comment by J3m5** [comment]
[When the whitelist was implemented](https://github.com/feathersjs-ecosystem/feathers-vuex/commit/9cd228405e15cf2f867cfdbc01a7174d897c6de0#diff-49ecb182ead299d15579b5ecb4ff549dR24), it was not added in the filter.
So we need to know if this filtering is really necessary, if yes add the whitelist to the filtering and if not remove the whole filtering.

**Comment by kshitizshankar** [comment]
> I think it's because of this:
> 
> ```js
> const customOperators = Object.keys(q).filter(
>   k => k[0] === '$' && !defaultOps.includes(k)
> )
> const cleanQuery = _omit(q, customOperators)
> ```
> 
> The filtering doesn't take the whitelist into account, I've seen it before but I don't know why it's done because the `cleanQuery` function is there for that and it is recursive.

Yeah I get what is happening in that flow - just don't understand why 😅 
So I have no idea if this is a bug or if there are other concerns that led to removing all '$' operators from the top-level of the query except those specifically hard-coded in there. 

Hopefully, someone else can shed some light on this!

**Comment by J3m5** [comment]
I think that we can safely remove this code block because[ the `filterQuery` function doesn't keep the operator if it's not in the given operators list](https://github.com/feathersjs/feathers/blob/3215714766f047cc9526f260796467b3c52f8d15/packages/adapter-commons/src/filter-query.ts#L47).

**Comment by kshitizshankar** [comment]
@J3m5 Oh perfect - This would make my life a lot easier for sure if it can be safely removed!

**Comment by marshallswain** [comment]
I would really appreciate a PR if this will benefit users.  I'm a bit strapped for time for the next week or so, but I can review PRs and do releases.  Thanks guys!

**Comment by marshallswain** [comment]
@kshitizshankar thanks for the details in your last comment.  I was able to make the PR and have published it as `feathers-vuex@3.12.1`.  Please reopen this issue if you still find that there's a problem.
